Extend kinematics for reporting xyz-axis min/max

mcmatrix opened the pull request:
Extend kinematic classes for reporting XYZ-axis min/max values.
These values can be used in the movement menu.

Example output:
```
printer['toolhead'].max_accel = 2000.0
printer['toolhead'].homed_axes =
printer['toolhead'].axis_max = {'y': 128.52, 'x': 128.52, 'z': 231.47}
printer['toolhead'].estimated_print_time = 3157.17333085
printer['toolhead'].max_velocity = 500.0
printer['toolhead'].print_time = 0.001
printer['toolhead'].axis_min = {'y': -128.52, 'x': -128.52, 'z': -5.0}
printer['toolhead'].max_accel_to_decel = 1000.0
printer['toolhead'].position = Coord(x=0.0, y=0.0, z=0.0, e=0.0)
```

mcmatrix commented:
Hi @KevinOConnor 
This is my attempt to extend reporting in kinematic classes.
As an example, I added it only for delta. Is this approach okay for you?
I probably need help to get correct values for other printer types (I have only delta live printer to test with).

mcmatrix commented:
Cartesian limites are here?
```
x -> self.limits[0][0], self.limits[0][1]
y -> self.limits[1][0], self.limits[1][1]
z -> ?
```

KevinOConnor commented:
Thanks.  At a high-level it looks fine to me.  A couple of minor comments:
- I think it would be preferable to use full words in the exported names (eg, `axis_minimum`).
- I'd avoid `round()` and export the actual value.
- For delta, I think the z max is `self.max_z`.
- For cartesian (and corexy), the limit is in `rail.get_range()`.
- If we add any exported variables, the docs/Command_Templates.md file should also be updated.

-Kevin

mcmatrix commented:
Thank you for the info @KevinOConnor 
Implemented reporting in all kinematic classes.
Only in `winch.py`, I wasn't sure, added just bounds from -200 to 200

KevinOConnor commented on klippy/kinematics/cartesian.py:
```diff
@@ -118,7 +118,15 @@ def check_move(self, move):
     def get_status(self, eventtime):
         axes = [a for a, (l, h) in zip("xyz", self.limits) if l <= h]
-        return { 'homed_axes': "".join(axes) }
+        axes_min = {}
+        axes_max = {}
```
It would be preferable to export a `homing.Coord()` instead of a dictionary so that it matches the format of other exported coordinates.  For example, see how `toolhead.get_status()` exports the `'position'` parameter.

KevinOConnor commented on klippy/kinematics/delta.py:
```diff
@@ -146,7 +146,17 @@ def check_move(self, move):
-        return {'homed_axes': '' if self.need_home else 'xyz'}
+        return {
+            'homed_axes': '' if self.need_home else 'xyz',
+            'axis_minimum': {
+                'x': -math.sqrt(self.max_xy2),
+                'y': -math.sqrt(self.max_xy2),
+                'z': self.min_z},
+            'axis_maximum': {
+                'x': math.sqrt(self.max_xy2),
+                'y': math.sqrt(self.max_xy2),
+                'z': self.max_z}
```
It would be preferable to calculate `sqrt(self.max_xy2)` once and then use that calculated value 4 times.

KevinOConnor commented on klippy/kinematics/polar.py:
```diff
@@ -108,7 +108,17 @@ def check_move(self, move):
+        return {
+            'homed_axes': xy_home + z_home,
+            'axis_minimum': {
+                'x': -math.sqrt(self.limit_xy2),
+                'y': -math.sqrt(self.limit_xy2),
+                'z': self.limit_z[0]},
+            'axis_maximum': {
+                'x': math.sqrt(self.limit_xy2),
+                'y': math.sqrt(self.limit_xy2),
+                'z': self.limit_z[1]}
```
The limits on polar should also come from `self.rails[0].get_range()` and `self.rails[1].get_range()`.

KevinOConnor reviewed (commented):
Thanks.  Looks good, but I did see a couple of minor things.

-Kevin

KevinOConnor commented on klippy/kinematics/winch.py:
```diff
@@ -47,7 +47,11 @@ def check_move(self, move):
         # XXX - homed_checks and rail limits not implemented
-        return {'homed_axes': 'xyz'}
+        return {
+            'homed_axes': 'xyz',
+            'axis_minimum': {'x': -200, 'y': -200, 'z': -200},
+            'axis_maximum': {'x': 200, 'y': 200, 'z': 200}
```
There's no enforcement of boundaries on winch.  However, a good approximation would be to limit to the minimum/maximum x,y,z of `self.anchors`.  For example, `min_x = min([a[0] for a in self.anchors])`.

-Kevin

mcmatrix commented:
Thanks @KevinOConnor 
Implemented all your suggestions.
Hopefully got all right this time.

KevinOConnor commented:
Thanks.

-Kevin
